lib/pos_remote_node_indexer.go

```diff
@@ -0,0 +1,231 @@
+package lib
+
+import "github.com/deso-protocol/core/bls"
+
+// RNIndexId is a custom string type used for identifying different types of remote node indices.
+type RNIndexId string
+
+// Constants for different types of remote node indices.
+const (
+	RNIndexId_Validator              RNIndexId = "VALIDATOR"
+	RNIndexId_ValidatorAttempted     RNIndexId = "VALIDATOR_ATTEMPTED"
+	RNIndexId_NonValidator_Outbound  RNIndexId = "NONVALIDATOR_OUTBOUND"
+	RNIndexId_NonValidator_Inbound   RNIndexId = "NONVALIDATOR_INBOUND"
+	RNIndexId_NonValidator_Attempted RNIndexId = "NONVALIDATOR_ATTEMPTED"
+)
+
+// RemoteNodeIndexer is a structure that holds information about all remote nodes and their indices.
+type RemoteNodeIndexer struct {
+	// AllRemoteNodes is a map storing all remote nodes by their IDs.
+	AllRemoteNodes map[RemoteNodeId]*RemoteNode
+
+	// Indices for various types of remote nodes.
+	ValidatorIndex             *RemoteNodeIndex[bls.PublicKey]
+	ValidatorAttemptedIndex    *RemoteNodeIndex[bls.PublicKey]
```

Comment:
Just for my understanding, why do we need separate maps for connected validators and attempted validators?

We store the validator's connected status in the RemoteNode itself, so it's easy for us to look up the validator by public key and then check its status no? If so, then it's an O(1) operation to perform such a lookup. We don't need to separate ValidatorIndex and ValidatorAttemptedIndex, and can instead have a single maps for them.